src/lib.rs

```diff
@@ -0,0 +1,117 @@
+        .build_inputs
+        .par_iter()
+        .map(|build_input| {
+            // check if the build_input has already be processed
+            let done = {
+                let mut collected_paths = collected_paths.lock().unwrap();
+                match &build_input.output_path {
+                    Some(output_path) => !collected_paths.insert(output_path.clone()),
+                    None => false,
+                }
+            };
+
+            if done {
+                log::debug!(
+                    "Skipping already processed derivation: {}",
+                    build_input.attribute_path
+                );
+                Vec::new()
+            } else {
+                process(
+                    collected_paths,
+                    flake_ref,
+                    system,
+                    &build_input.attribute_path,
+                    offline,
+                    lib,
+                )
+            }
+        })
+        .flatten()
+        .collect();
+
+    // combine the children and this derivation into a single Vec
+    let mut descriptions = children;
+    descriptions.push(description);
+    descriptions
+}
+
+pub fn nixtract(
+    flake_ref: impl AsRef<str>,
+    system: Option<impl AsRef<str>>,
+    attribute_path: Option<impl AsRef<str>>,
+    offline: &bool,
+) -> Result<Vec<DerivationDescription>> {
+    let flake_ref = flake_ref.as_ref();
+    // Convert system to a Option<&str>
+    let system = system.as_ref().map(AsRef::as_ref);
+    let attribute_path = attribute_path.as_ref().map(AsRef::as_ref);
+
+    // Writes the `lib.nix` file to the tempdir and stores its path
+    let lib = nix::lib::Lib::new()?;
+
+    log::info!(
+        "Starting nixtract with flake_ref: {}, system: {}, attribute_path: {:?}",
+        flake_ref,
+        system
+            .map(AsRef::as_ref)
+            .unwrap_or("builtins.currentSystem"),
+        attribute_path.map(AsRef::as_ref).unwrap_or("")
+    );
+
+    let collected_paths: Arc<Mutex<std::collections::HashSet<String>>> =
+        Arc::new(Mutex::new(std::collections::HashSet::new()));
+
+    // call find_attribute_paths to get the initial set of derivations
+    let attribute_paths =
+        nix::find_attribute_paths(flake_ref, system, attribute_path, offline, &lib)?;
+
+    // Combine all AttributePaths into a single Vec
+    let mut derivations: Vec<FoundDrv> = Vec::new();
+    for attribute_path in attribute_paths {
+        derivations.extend(attribute_path.found_drvs);
+    }
+
+    // call process on every foundDrv
+    let descriptions: Vec<DerivationDescription> = derivations
+        .par_iter()
+        .map(|found_drv| {
+            process(
+                &collected_paths,
+                flake_ref,
+                system,
+                &found_drv.attribute_path,
+                offline,
+                &lib,
+            )
+        })
+        .flatten()
+        .collect();
```

Comment:
It looks to me like this implementation can describe many times the same derivation, whereas the current (Python) one does not (using a mutable set to check if a derivation has already been described).

Reply:
I don't think it does: https://github.com/tweag/nixtract/pull/26/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R29